### PR TITLE
feat(registry): add repo registry — machine-readable module/file/owner/criticality map

### PR DESCRIPTION
**Pull request by diberry**
- [x] Explore codebase and understand current state
- [x] Fix `parsePackages()` to sort directory entries (determinism)
- [x] Fix `resolveModuleToPackage()` to handle package-qualified paths (`packages/squad-sdk/src/`, `packages/squad-cli/src/`)
- [x] Fix `resolveFileToModule()` to derive package from stripped file-path prefix (returns correct `squad-sdk` vs `squad-cli` for SDK files)
- [x] Add `--out <path>` flag to generator script (enables test isolation)
- [x] Fix registry-sync.test.ts to write to temp file instead of overwriting committed file
- [x] Fix registry-sync.test.ts to compare parsed JSON objects (avoids EOL false positives on Windows)
- [x] Fix misleading test description ("scaffolding" → "medium or unknown tier")
- [x] Regenerate `.squad/repo-registry.json` with all fixes applied
- [x] All 40 registry tests pass

### CODE REVIEW

**Comment by diberry** [comment]
@copilot Please review this PR. Focus on: Is the routing.md parser robust? Are the utility functions correct? Is the generated JSON schema complete? Any edge cases in resolveFileToModule()?

**Comment by diberry** [comment]
## 🏗️ Architecture Review: Repo Registry (PR #737)

**STATUS: ✅ APPROVE**

### Summary
PR #737 implements a robust, machine-readable repository registry that correctly captures module ownership, criticality tiers, and file mappings. The design is sound, extensible, and ready to support downstream consumers (#670, #733).

### Strengths
- **Schema correctness**: Registry structure properly normalized with clear top-level keys (modules, criticalFiles, packages, templateFiles, metadata)
- **Source integrity**: SHA256 hashes for routing.md, templates.ts, and package.json enable change detection in CI
- **Complete module coverage**: All 19 modules from routing.md correctly parsed and stored with primary + optional secondary owners
- **Utility API solid**: Six well-documented query functions (loadRegistry, resolveFileToModule, getModuleOwner, getFileCriticality, getPackageForFile, getModulesByOwner) provide clean downstream interfaces
- **Fallback logic clear**: Unknown files → Flight ownership; unknown criticality → impact:0; lazy caching prevents repeated I/O
- **Test guards effective**: Freshness sync test catches source drift; normalize-timestamp prevents false positives; full JSON validation included
- **Forward-compatible**: Package prefix normalization and longest-prefix matching support both #670 (file layout manifest) and #733 (PR impact analysis) without breaking changes

### Notes
- Registry generation script robustly handles malformed routing.md with explicit error messages—excellent for catch-and-fix workflow
- Template extraction correctly identifies 29 entries across 4 tiers (critical, high, medium); CI workflows and squad.agent.md properly marked as critical
- All module entries include package mapping or null for root-level tests

### Recommendation
**Ready to merge.** Meets all architectural criteria for module registry. Ready for #670 and #733 to import and consume.

**Comment by diberry** [comment]
## Test Coverage Review - PR #737 (Registry Sync)

### Test Execution Summary
- **All tests passing**: 40 tests across 2 test files
- **Test files**: registry-sync.test.ts (4 tests), registry-utils.test.ts (36 tests)
- **Execution time**: 1.63s
- **Exit code**: 0

### Timestamp Normalization - SOLID
- `normalizeTimestamp()` helper correctly replaces generated timestamps with placeholder
- Regex pattern `/"generated"\s*:\s*"[^"]*"/` matches with flexible whitespace handling
- Both committed and freshly generated registries normalized before comparison
- Prevents false failures from timestamp drift

### Edge Cases Covered

**Unknown paths:**
- README.md resolves to `unknown` module with `Flight` primary
- .github/workflows/squad-ci.yml resolves to `unknown` (outside module table)
- Unrecognized module keys default to `Flight` primary

**Module resolution robustness:**
- Longest-prefix matching for nested paths (e.g., src/cli/shell/components/)
- Trailing slash consistency across path variants
- File-level entries (e.g., src/index.ts) vs. directory paths
- Windows vs. Unix path normalization

**Malformed routing.md:**
- Script throws explicit error if "## Module Ownership" section missing
- Validates minimum row format (modulePath + primary owner required)
- Normalizes emoji characters before parsing

**Package discovery:**
- squad-sdk and squad-cli correctly identified
- Root-level files return null for package lookup
- .github files return null for package

### Freshness Test is Robust
- **Pattern**: Regenerate -> read committed -> normalize both -> deep-compare
- **Sync guard**: Fails if source files (routing.md, templates.ts, package.json) edited without running npm run registry
- **CI integration**: Detects drift before merge
- **Script exit check**: Validates generator returns code 0
- **JSON validation**: Confirms committed registry is parseable
- **Required keys check**: Verifies all top-level properties exist

### Query API Coverage
- **resolveFileToModule()**: 8 scenarios (known modules, nested specificity, unknown paths, trailing slashes, file-level entries)
- **getModuleOwner()**: 5 scenarios (known/unknown modules, primary/secondary fallbacks, null secondary handling)
- **getFileCriticality()**: 6 scenarios (critical files, high tier, default fallbacks)
- **getPackageForFile()**: 5 scenarios (workspace packages, root/.github files returning null)
- **getModulesByOwner()**: 5 scenarios (multi-module owners, single-module owners, empty results)
- **loadRegistry()**: 7 scenarios (top-level structure, module integrity, hash metadata)

### Minor Observations (Not blockers)
1. **Criticality fallback**: getFileCriticality() returns 'medium' for any file in known modules; consider if this should distinguish between unknown modules ('unknown') vs. unmapped files in known modules ('medium')
2. **Hash verification not in tests**: Registry metadata hashes are generated but not validated in tests - consider adding a test to verify hashes regenerate consistently
3. **Secondary owner can be null**: Tests verify this, but might warrant defensive checks in downstream consumers

---

### VERDICT: APPROVE

**Rationale:**
- All 40 tests passing with 100% execution success
- Timestamp normalization prevents flaky CI checks
- Comprehensive edge case coverage for unknown/malformed input
- Query API well-tested against real registry data
- Freshness guard properly detects source file drift
- Clear error messages guide developers to fix violations

**Ready to merge.** Test suite is production-ready and provides strong guardrails for registry freshness and drift detection.

**Comment by diberry** [comment]
## CI/Process Review — PR #737 (Repo Registry)

### Status: ✅ APPROVE

All process and CI requirements met. Branch is clean and ready to merge.

### Verification Checklist

#### ✅ Commit Structure
- **Count:** Exactly 1 commit on branch (dev..squad/736-repo-registry)
- **Message:** Clear semantic commit: \eat(registry): add repo registry — machine-readable module/file/owner/criticality map\
- **Co-authored:** Includes Copilot trailer ✓

#### ✅ CHANGELOG Entry
- **Format:** Follows existing pattern under \## [Unreleased]\ section
- **Location:** Top of Unreleased, under \### Added — Repo Registry (#736)\
- **Content:** Well-structured with:
  - Repo registry definition + file path
  - Scripts added: \scripts/generate-repo-registry.mjs\ (generator) + \scripts/registry-utils.mjs\ (utils)
  - NPM script \
pm run registry\ documented
  - CI freshness guard pattern noted
  - Issue references: Unblocks #670 and #733 ✓

#### ✅ Package.json Updates
- **New script added:** \"registry": "node scripts/generate-repo-registry.mjs"\ (line 21)
- **Location:** Correct alphabetical order among dev scripts
- **No version bump:** version remains \

File: test/registry-sync.test.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+    // 2. Regenerate by running the generator script
+    execSync(`node ${GENERATOR_SCRIPT}`, {
+      cwd: ROOT,
+      encoding: 'utf-8',
+      timeout: 30_000,
+    });
+
+    // 3. Read the freshly generated registry
+    const freshContent = readFile(REGISTRY_PATH);
```
[the remark this answers is not in the document]
Comment (diberry) [reply test/registry-sync.test.ts]: Fixed: Test now generates to a temp file via CLI arg, compares against committed copy, and cleans up in a finally block. Working tree stays clean.

File: scripts/generate-repo-registry.mjs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+  for (const pattern of pkgJson.workspaces || []) {
+    const baseDir = join(ROOT, pattern.replace('/*', ''));
+    for (const entry of readdirSync(baseDir, { withFileTypes: true })) {
+      if (entry.isDirectory()) {
+        packages[entry.name] = { path: `${pattern.replace('/*', '')}/${entry.name}` };
+      }
+    }
+  }
+
```
[the remark this answers is not in the document]
Comment (diberry) [reply scripts/generate-repo-registry.mjs]: Fixed: Sorted package keys alphabetically before writing JSON — output is now deterministic regardless of filesystem readdir order.

File: scripts/generate-repo-registry.mjs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+  console.log(`  ✅ Found ${Object.keys(packages).length} workspace packages`);
+  return packages;
+}
+
+// ─── 4. Helpers ────────────────────────────────────────────────────────
+
+function resolveModuleToPackage(modulePath) {
+  if (modulePath.startsWith('src/')) return 'squad-cli';
```
[the remark this answers is not in the document]
Comment (diberry) [reply scripts/generate-repo-registry.mjs]: Fixed: resolveModuleToPackage() now checks which package directory actually contains the module path using existsSync(), instead of hard-coding squad-cli for all src/* paths.

File: scripts/registry-utils.mjs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  }
+
+  // Sort by path length descending so "src/cli/shell/components/" matches before "src/cli/"
+  const sortedModules = Object.keys(registry.modules)
+    .sort((a, b) => b.length - a.length);
+
+  for (const modulePath of sortedModules) {
+    if (normalized.startsWith(modulePath) || normalized === modulePath.replace(/\/$/, '')) {
+      return { module: modulePath, ...registry.modules[modulePath] };
+    }
```
[the remark this answers is not in the document]
Comment (diberry) [reply scripts/registry-utils.mjs]: Fixed: resolveFileToModule() now determines the package from the file path prefix (packages/squad-sdk/ → squad-sdk) during prefix stripping, rather than relying on the module table's package field.

File: test/registry-utils.test.ts [hunk 5]
```diff
@@ -0,0 +1,309 @@
...
+  it('returns "high" for .squad/routing.md', () => {
+    expect(getFileCriticality('.squad/routing.md')).toBe('high');
+  });
+
+  it('returns "scaffolding" (default) for an unknown/unmapped file', () => {
+    const result = getFileCriticality('src/some-random-file.ts');
+    // Files in known modules but not in criticalFiles get "medium"
+    // Files not matching any module get "unknown"
+    expect(['medium', 'unknown']).toContain(result);
+  });
```
[the remark this answers is not in the document]
Comment (diberry) [reply test/registry-utils.test.ts]: Fixed: Updated test description from "scaffolding" to "medium or unknown" to match the actual tier vocabulary (critical/high/medium/unknown) used in the implementation.

File: test/registry-sync.test.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    // 4. Normalize timestamps in both
+    const freshNormalized = normalizeTimestamp(freshContent.trim());
+    const committedNormalized = normalizeTimestamp(committedContent.trim());
+
+    // 5. Compare — if this fails, a source file was changed without
+    //    running `npm run registry` (or `node scripts/generate-repo-registry.mjs`)
+    expect(freshNormalized).toBe(committedNormalized);
+  });
```
[the remark this answers is not in the document]
Comment (diberry) [reply test/registry-sync.test.ts]: Fixed: Added CRLF→LF normalization (.replace(/\r\n/g, '\n')) before string comparison in the freshness test, so it passes on Windows with core.autocrlf=true.
